src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,155 @@
+                        return true;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        completion.SetCanceled();
+                    }
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doesn't affect the queue processing.
+                        completion.SetException(exception);
+                    }
+
+                    // Tell the queue to ignore any mutations from this request
+                    return false;
+                }, requestCancellationToken);
+            _queue.Enqueue(item);
+
+            return completion.Task;
+        }
+
+        private async Task ProcessQueueAsync()
+        {
+            // Keep track of solution state modifications made by LSP requests
+            Solution? lastMutatedSolution = null;
+            var queueToken = _cancelSource.Token;
+
+            while (!queueToken.IsCancellationRequested)
+            {
+                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+
+                // Create a linked cancellation token to cancel any requests in progress when this shuts down
+                var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(queueToken, work.CancellationToken).Token;
+
+                // The "current" solution can be updated by non-LSP actions, so we need it, but we also need
+                // to merge in the changes from any mutations that have been applied to open documents
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, lastMutatedSolution);
+
+                Solution? mutatedSolution = null;
+                var context = new RequestContext(solution, s => mutatedSolution = s, work.ClientCapabilities, work.ClientName);
+
+                if (work.MutatesSolutionState)
+                {
+                    // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                    var ranToCompletion = await work.Callback(context, cancellationToken).ConfigureAwait(false);
+
+                    // If the handling of the request failed, the exception will bubble back up to the caller, but we
+                    // still need to react to it here by throwing away solution updates
+                    if (ranToCompletion)
+                    {
+                        lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
+                    }
+                }
+                else
+                {
+                    // Non mutating request get given the current solution state, but are otherwise fire-and-forget
+                    _ = work.Callback(context, cancellationToken);
```

Comment:
fire-and-forget means we lose exceptions.  this seems non-ideal.

Reply:
agree - should definitely track these in some manner, maybe NFW?

Reply:
Will do this with the failed mutation handling. Since only an exception inside the callback function (ie, not from the actual request handler) would possibly bubble out of here, shutting down the server is probably the only course of action really.